presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFilterManager.java

```diff
@@ -0,0 +1,319 @@
+                partitionEndOffsets = calcTopicPartitionEndOffsetMap(partitionEndOffsets,
+                        p -> (range.getEnd() >= 0) ? Optional.of(range.getEnd()) : Optional.empty());
+            }
+
+            // push down timestamp if possible
+            if (offsetTimestampRanged.isPresent()) {
+                try (KafkaConsumer<byte[], byte[]> kafkaConsumer = consumerFactory.create()) {
+                    Optional<Range> finalOffsetTimestampRanged = offsetTimestampRanged;
+                    partitionBeginOffsets = calcTopicPartitionBeginOffsetMap(partitionBeginOffsets,
+                            p -> findOffsetsByTimestamp(kafkaConsumer, p, finalOffsetTimestampRanged.get().getBegin()));
+                    if (isTimestampUpperBoundPushdownEnabled(session, kafkaTableHandle.getTopicName())) {
+                        partitionEndOffsets = calcTopicPartitionEndOffsetMap(partitionEndOffsets,
+                                p -> findOffsetsByTimestamp(kafkaConsumer, p, finalOffsetTimestampRanged.get().getEnd()));
+                    }
+                }
+            }
+
+            // push down partitions
+            final Set<Long> finalPartitionIdsFiltered = partitionIdsFiltered;
+            List<PartitionInfo> partitionFilteredInfos = partitionInfos.stream()
+                    .filter(partitionInfo -> finalPartitionIdsFiltered.contains((long) partitionInfo.partition()))
+                    .collect(toImmutableList());
+            return new KafkaFilteringResult(partitionFilteredInfos, partitionBeginOffsets, partitionEndOffsets);
+        }
+        return new KafkaFilteringResult(partitionInfos, partitionBeginOffsets, partitionEndOffsets);
+    }
+
+    private boolean isTimestampUpperBoundPushdownEnabled(ConnectorSession session, String topic)
+    {
+        try (AdminClient adminClient = adminFactory.create()) {
+            ConfigResource topicResource = new ConfigResource(ConfigResource.Type.TOPIC, topic);
+
+            DescribeConfigsResult describeResult = adminClient.describeConfigs(Collections.singleton(topicResource));
+            Map<ConfigResource, Config> config = describeResult.all().get();
+
+            if (config != null) {
+                Config c = config.get(topicResource);
+                String timestampType = c.get(TOPIC_CONFIG_TIMESTAMP_KEY).value();
+                if (TOPIC_CONFIG_TIMESTAMP_VALUE_LOG_APPEND_TIME.equals(timestampType)) {
+                    return true;
+                }
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(KAFKA_SPLIT_ERROR, format("Failed to get topic(%s) config", topic), e);
+        }
+        return KafkaSessionProperties.isTimestampUpperBoundPushdownEnabled(session);
+    }
+
+    private static Optional<Long> findOffsetsByTimestamp(KafkaConsumer<byte[], byte[]> kafkaConsumer, TopicPartition topicPartition, long timestamp)
+    {
+        try {
+            if ((timestamp + "").length() == 16) {
```

Comment:
This is super cryptic. Can you explain in a comment what is happening here?

Reply:
when I rebase code from latest. It seems the timestamp precision become 16(us) other than 13(ms)